numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -1927,17 +1927,26 @@ PyArray_Zero(PyArrayObject *arr)
         return NULL;
     }
 
-    obj=PyInt_FromLong((long) 0);
+    if (zero_obj == NULL) {
+        zero_obj = PyInt_FromLong((long) 0);
+        if (zero_obj == NULL) {
+            return NULL;
+        }
+    }
     if (PyArray_ISOBJECT(arr)) {
```

Comment:
I'm not sure I follow. AFAIK ``PyArray_DESCR(arr)->f->setitem(one_obj, oneval, arr)`` will assume ``oneval`` is a valid pointer to a block of memory large enough to hold sizeof(PyIntObject), but it is actually a ``void* NULL`` at this point, with no valid content. So ``setitem`` will effectively call  ``memcpy(oneval, one_obj, sizeof(PyIntObject)`` which will attempt to copy around 32 bytes into a NULL pointer.